File: View Controllers/SettingsViewController.m
//
//  SettingsViewController.m
//  aestheticodes
//
//  Created by Kevin Glover on 03/07/2014.
//  Copyright (c) 2014 Horizon. All rights reserved.
//

#import "Experience.h"
#import "SettingsViewController.h"

@interface SettingsViewController ()

@end

@implementation SettingsViewController

- (void)viewDidLoad
{
	[super viewDidLoad];
	[self loadValues];
}


-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
	UITableViewCell* cell = [tableView cellForRowAtIndexPath:indexPath];
	UIView* view = [cell contentView];
	for (UIView* subview in [view subviews])
	{
		if([subview isKindOfClass:[UITextField class]])
		{
			[subview becomeFirstResponder];
		}
	}
}

-(void)loadValues
{
	self.title = [NSString stringWithFormat:@"%@ Settings", self.experience.name];
	
	minRegions.text = [NSString stringWithFormat:@"%d", self.experience.minRegions];
	maxRegions.text = [NSString stringWithFormat:@"%d", self.experience.maxRegions];
	maxRegionValue.text = [NSString stringWithFormat:@"%d", self.experience.maxRegionValue];
	maxEmptyRegions.text = [NSString stringWithFormat:@"%d", self.experience.maxEmptyRegions];

	validationRegions.text = [NSString stringWithFormat:@"%d", self.experience.validationRegions];
	validationRegionValue.text = [NSString stringWithFormat:@"%d", self.experience.validationRegionValue];
	checksum.text = [NSString stringWithFormat:@"%d", self.experience.checksumModulo];
}

-(void)viewWillDisappear:(BOOL)animated
{
	[super viewWillDisappear:animated];
	[self saveValues];
}

-(IBAction)validate:(UITextField*)sender
{
	[self isValid:sender];
}

-(BOOL)isValid:(UITextField*)sender
{
	bool valid = true;
	if(![self isInt:sender.text])
	{
		valid = false;
	}
	NSInteger value = [sender.text integerValue];
	if(value < 0 || value > 20)
	{
		valid = false;
	}
	
	if(sender == minRegions || sender == maxRegions)
	{
		if([self isInt:minRegions.text] && [self isInt:maxRegions.text])
		{
			NSInteger minValue = [minRegions.text integerValue];
			NSInteger maxValue = [maxRegions.text integerValue];
			if(minValue > maxValue)
			{
				valid = false;
			}
		}
	}
	
	if(valid)
	{
		sender.rightViewMode = UITextFieldViewModeNever;
	}
	else
	{
		sender.rightView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"error.png"]];
		sender.rightViewMode = UITextFieldViewModeAlways;
	}
	
	return valid;
}

-(bool)isInt:(NSString*)string
{
	NSInteger value = [string integerValue];
	if(value != 0)
	{
		return true;
	}
	return [string isEqualToString:@"0"];
}

-(void)saveValue:(UITextField*)sender
{
	if([self isValid:sender])
	{
		[self.experience setIntValue:[sender.text integerValue] key:sender.restorationIdentifier];
	}
}

-(void)saveValues
{
	[self saveValue:minRegions];
	[self saveValue:maxRegions];
	[self saveValue:maxRegionValue];
	[self saveValue:maxEmptyRegions];

	[self saveValue:validationRegions];
	[self saveValue:validationRegionValue];
	
	[self saveValue:checksum];	
}

@end

File: View Controllers/SettingsViewController.swift
//
//  ACSettingsViewController.swift
//  aestheticodes
//
//  Created by Kevin Glover on 09/06/2014.
//  Copyright (c) 2014 horizon. All rights reserved.
//

import Foundation
import UIKit

class SettingsViewController: UITableViewController, UITextFieldDelegate
{
	@IBOutlet var minRegions: UITextField
	@IBOutlet var maxRegions: UITextField
	@IBOutlet var maxRegionValue: UITextField
	@IBOutlet var maxEmptyRegions: UITextField

	@IBOutlet var validationRegions: UITextField
	@IBOutlet var validationRegionValue: UITextField
	@IBOutlet var checksum: UITextField
	
	override func viewDidLoad()
	{
		loadValues()
	}
	
	override func tableView(tableView: UITableView!, didSelectRowAtIndexPath indexPath: NSIndexPath!)
	{
		let cell = tableView.cellForRowAtIndexPath(indexPath)
		let view = cell.contentView
		for subview : AnyObject in view.subviews
		{
			if subview is UITextField
			{
				subview.becomeFirstResponder()
			}
		}
	}
	
	func loadValues()
	{
		NSLog("\(markerSettings.minRegions)")
		minRegions.text = "\(markerSettings.minRegions)"
		maxRegions.text = "\(markerSettings.maxRegions)"
		maxRegionValue.text = "\(markerSettings.maxRegionValue)"
		maxEmptyRegions.text = "\(markerSettings.maxEmptyRegions)"
		
		validationRegions.text = "\(markerSettings.validationRegions)"
		validationRegionValue.text = "\(markerSettings.validationRegionValue)"
		
		checksum.text = "\(markerSettings.checksumModulo)"
	}
	
	override func viewWillDisappear(animated: Bool)
	{
		super.viewWillDisappear(animated)
		saveValues()
	}
	
	@IBAction func validate(sender : UITextField)
	{
		isValid(sender)
	}
	
	func isValid(sender: UITextField) -> Bool
	{
		var valid = true
		let value = sender.text.toInt()
		if !value || value > 20
		{
			valid = false
		}
		
		if sender == minRegions || sender == maxRegions
		{
			if minRegions.text.toInt() && maxRegions.text.toInt()
			{
				if minRegions.text.toInt() > maxRegions.text.toInt()
				{
					valid = false
				}
			}
		}
		
		if valid
		{
			sender.rightViewMode = UITextFieldViewMode.Never
		}
		else
		{
			sender.rightView = UIImageView(image: UIImage(named: "error.png"))
			sender.rightViewMode = UITextFieldViewMode.Always
		}
		
		return valid
	}
	
	func saveValue(sender: UITextField)
	{
		if isValid(sender)
		{
			markerSettings.setIntValue(sender.text.toInt(), key: sender.restorationIdentifier)
		}
	}
	
	func saveValues()
	{
		saveValue(minRegions)
		saveValue(maxRegions)
		saveValue(maxRegionValue)
		saveValue(maxEmptyRegions)
		
		saveValue(validationRegions)
		saveValue(validationRegionValue)
		
		saveValue(checksum)
	}
}